s3store/s3.go

```diff
@@ -0,0 +1,47 @@
+package s3store
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// Store uses s3 as backing store
+type Store struct {
+	Bucket string
+}
+
+// Upload fulfills attache.Store interface
+func (s Store) Upload(file *bytes.Reader, fileType string) (string, error) {
+	fileName := filename(fileType)
+	filePath := fmt.Sprintf("https://s3-%s.amazonaws.com/%s/%s", os.Getenv("AWS_REGION"), s.Bucket, fileName)
```

Comment:
👍 indeed full URL will make it easier to wield on the app side.

Can consider deferring to env to allow templating of this string; avoid bucket name / region issues (i recall there is some awkwardness). So if app need to configure a specific cloudfront/human friendly domain they can setup their own without too much hassle

```
FILE_URL=https://cdn.example.com/{{ .Bucket }}/{{ .Filename }}
```

Reply:
actually, i take that back. the return string should NOT be the s3 url (or google cloud storage url, or azure storage url, ...)

- after a jpg file is uploaded to attache, it will store the file as unique s3 object, e.g. "87654321.jpg"
- attache will return the `{ "path": "...", ... }` json; your app will store the necessary value in db

so far so good, right?

But your app should never render `<img src=...>` using an s3 url. Instead your app should always point to attache 
- i.e. `<img src=http://attache.example.com/87654321.jpg>`
- or `<img src=http://attache.example.com/resize/100x/87654321.jpg>`

So, why give that s3 url for your app to store?

And your app likely want to front attache GET requests with whatever cdn / varnish / custom domain, e.g. `https://random.cloudfront.net`...

And also use different processing instructions on different landing pages, e.g. `/resize/100x`

So it is really simpler (and more correct) for your app to concern itself with the url prefix of the s3 object key `87654321.jpg`, e.g. `https://random.cloudfront.net/87654321.jpg` or `https://random.cloudfront.net/resize/100x/87654321.jpg`

So, the `Store#Upload` function should only return the "filename" (aka s3 key)

Reply:
I had the same thought too but was thinking that if we don't store the full URL in the DB, then what's in the DB is considered incomplete? i.e. you'll need the ENV of region and bucket in the app itself to piece together the information of where the image is.
(I was looking at Uploadcare as well, and they return the full URL.)

Also, is there a the possibility that the DOWNLOAD/RESIZE portion is totally divorced as a another service? So it can accept image from anywhere?

Reply:
Working on the changes now btw.

Reply:
> what's in the DB is considered incomplete? i.e. you'll need the ENV of region and bucket in the app itself to piece together the information of where the image is.

No -- you're still thinking the "app" has to worry about the env, region, bucket.. it doesn't. the app only points to attache.

attache is the one configured with bucket, region, etc (see `s3store.Store` struct)

Reply:
Gotcha. Let me fix in the new PR.

Reply:
I was summoned :)

In fact Uploadcare widget returns both the URL (and users can choose to store it in the app for simplicity, and it covers 80% of use cases) and file metadata (via JS API.)

Also, knowing the URL, users can get all the [file metadata](https://uploadcare.com/documentation/rest/#file) later save/use some subset of it.